model/lists/Service/ListUpdater.php

```diff
@@ -0,0 +1,204 @@
+            $listClass->setLabel($payload['label']);
+        }
+
+        unset($payload['uri'], $payload['label']);
+
+        if (count($payload) > 500) {
+            throw new BadFunctionCallException('Payload contains too many items');
+        }
+
+        if (!$this->setListElements($listClass, $payload)) {
+            throw new RuntimeException('Error saving list items');
+        }
+    }
+
+    /**
+     * Updates the ValueCollection instance corresponding to a class.
+     *
+     * The payload is used to call ValueCollectionService::persist(): Depending
+     * on the particular underlying repository class used (instance of
+     * ValueCollectionRepositoryInterface), that may cause removing all existing
+     * items first (i.e. for remote lists) or merging the values provided with
+     * pre-existing values.
+     */
+    private function setListElements(
+        core_kernel_classes_Class $listClass,
+        array $payload
+    ): bool {
+        // This method retrieves only elements corresponding to the URIs that
+        // are modified by the request (i.e. present in the POST data) instead
+        // of all list items.
+        //
+        // Retrieving existing elements is needed in order to return an accurate
+        // value for Value::hasModifiedUri() calls, as the repository might
+        // depend on that to check if an element needs to be created or updated.
+        //
+        // Note also we cannot POST two items with the same former URI, so there
+        // is no need to check for duplicates in the input data itself.
+        //
+        $elements = $this->getElementsFromPayload($payload);
+        $collection = $this->valueCollectionService->findAll(
+            $this->getListSearchInput($listClass, $elements)
+        );
+
+        foreach ($elements as $uriKey => $value) {
+            $newUri = trim($payload["uri_{$uriKey}"] ?? '');
+            $this->addOneElement($collection, $uriKey, $value, $newUri);
+        }
+
+        return $this->valueCollectionService->persist($collection);
+    }
+
+    private function addOneElement(
+        ValueCollection $valueCollection,
+        string $uri,
+        string $value,
+        string $newUri
+    ): void {
+        $element = $this->getValueByUriKey($valueCollection, $uri);
+
+        if ($element === null) {
+            $valueCollection->addValue(new Value(null, $newUri, $value));
+            return;
+        }
+
+        $element->setLabel($value);
+
+        if ($newUri) {
+            $element->setUri($newUri);
+        }
+    }
+
+    private function getValueByUriKey(
+        ValueCollection $valueCollection,
+        string $key
+    ): ?Value {
+        $uri = $this->getElementURIFromKey($key);
+        if (empty($uri)) {
+            return null;
+        }
+
+        return $valueCollection->extractValueByUri($uri);
+    }
+
+    private function getListSearchInput(
+        core_kernel_classes_Class $listClass,
+        $elements
```

Comment:
TODO 
- [ ] Add array typehint

```suggestion
        array $elements
```

Reply:
Added: 884d03b67116485aa9cb5098fd253be4913392a5